### PR TITLE
Reformat codebase and add `pre-commit`

### PR DESCRIPTION
**Pull request by zhijian-liu**
This PR reformats the codebase and integrates `pre-commit` for continuous integration. Contributors can first activate it locally by `pre-commit install`, which will then be triggered automatically at `git commit` to check, fix and reformat the commit.

### CODE REVIEW

**Review by kentangSJTU — APPROVED** [review]

**Review by CCInc — DISMISSED** [review]
Looks great! I'm curious, what's the advantage of doing it through pre-commit versus the old formatter workflow?

**Comment by zhijian-liu** [comment]
@CCInc, the advantages are mainly two-fold:
* It allows the contributors to perform the check locally before pushing the changes to the repo.
* It supports more than just reformatting the code: it also checks some type errors (see https://github.com/mit-han-lab/torchsparse/blob/dev/pre-commit/.pre-commit-config.yaml for more details).

**Review by sandeepnmenon — DISMISSED** [review]
This is great!

**Comment by zhijian-liu** [comment]
@CCInc @sandeepnmenon, this PR is ready for review. @kentangSJTU, I've tested the correctness of this version on semantic segmentation. It would be great if you could test it on object detection as well.

File: torchsparse/utils/utils.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+__all__ = ['make_ntuple']
+
+
+def make_ntuple(x: Union[int, List[int], Tuple[int, ...]],
+                ndim: int) -> Tuple[int, ...]:
+    if isinstance(x, int):
+        x = tuple(repeat(x, ndim))
+    elif isinstance(x, list):
+        x = tuple(x)
+
```
[the remark this answers is not in the document]
Comment (zhijian-liu) [reply torchsparse/utils/utils.py:17]: Fixed.

File: torchsparse/nn/modules/bev.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
+import math
```
[the remark this answers is not in the document]
Comment (zhijian-liu) [reply torchsparse/nn/modules/bev.py:1]: Fixed.

File: torchsparse/nn/functional/downsample.py [hunk 3]
```diff
@@ -1,70 +1,43 @@
...
+        xyz = coords[:, :3].unsqueeze(1).repeat(1, offsets.size(0), 1) + offsets
+        b = coords[:, 3:].repeat(1, offsets.size(0))
+        coords = torch.cat([xyz.view(-1, 3), b.view(-1, 1)], dim=1)
+
+        mask = ((coords[:, 0] % sample_stride[0] == 0)
+                & (coords[:, 1] % sample_stride[1] == 0)
+                & (coords[:, 2] % sample_stride[2] == 0))
+        coords = coords[mask]
+
+    coords = torch.unique(coords, dim=0)
```
[the remark this answers is not in the document]
Comment (zhijian-liu) [reply torchsparse/nn/functional/downsample.py:47]: That's a very good point! I have reordered `coords` to preserve this property. It would be great to add comments before this sort of design choice to let others understand the thinking process.

File: torchsparse/nn/modules/conv.py [hunk 4]
```diff
@@ -1,301 +1,72 @@
...
                  in_channels: int,
                  out_channels: int,
-                 kernel_size: Union[int, List[int], Tuple[int, int, int]] = 3,
-                 stride: Union[int, List[int], Tuple[int, int, int]] = 1,
+                 kernel_size: Union[int, Tuple[int, ...]] = 3,
+                 stride: Union[int, Tuple[int, ...]] = 1,
                  dilation: int = 1,
                  bias: bool = False,
-                 transpose: bool = False) -> None:
+                 transposed: bool = False) -> None:
```
[the remark this answers is not in the document]
Comment (CCInc) [reply torchsparse/nn/modules/conv.py:24]: I think it might break a lot of downstream packages

File: torchsparse/nn/functional/conv.py [hunk 5]
```diff
@@ -1,218 +1,147 @@
...
     def forward(ctx,
-                features,
-                kernel,
-                neighbor_map,
-                neighbor_offset,
+                input: torch.Tensor,
+                weight: torch.Tensor,
+                nbmaps: torch.Tensor,
+                nbsizes: torch.Tensor,
                 sizes,
```
Comment (CCInc) [inline torchsparse/nn/functional/conv.py:25]: type annotation here?

File: torchsparse/nn/modules/conv.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhijian-liu) [reply torchsparse/nn/modules/conv.py:24]: This change is to make our `nn.Conv3d` consistent with `torch.nn.ConvNd`. I will add a summary of API changes in the release note.

File: torchsparse/nn/functional/count.py [hunk 6]
```diff
@@ -1,18 +1,15 @@
...
-        if 'cuda' in str(idx.device):
-            outs = torchsparse_backend.count_forward(idx.contiguous(), num)
-        else:
-            outs = torchsparse_backend.cpu_count_forward(idx.contiguous(), num)
-        return outs
-
-
-def spcount(idx, num):
-    return CountGPU.apply(idx, num)
+def spcount(coords, num):
```
Comment (CCInc) [inline torchsparse/nn/functional/count.py:6]: type annotation

File: torchsparse/nn/functional/devoxelize.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        w[idx_query == -1] = 0
+        w /= w.sum(0) + 1e-8
+    return w
+
+
+class DevoxelizeFunction(Function):
+
+    @staticmethod
+    @custom_fwd(cast_inputs=torch.half)
+    def forward(ctx, feats, indices, weights):
```
Comment (CCInc) [inline torchsparse/nn/functional/devoxelize.py:57]: same

File: torchsparse/nn/functional/hash.py [hunk 8]
```diff
@@ -1,41 +1,37 @@
...
 
+    if offsets is not None:
+        assert offsets.dtype == torch.int, offsets.dtype
+        assert offsets.ndim == 2 and offsets.shape[1] == 3, offsets.shape
+        offsets = offsets.contiguous()
 
-def sphash(idx, koffset=None):
-    if koffset is None:
-        return HashGPU.apply(idx)
+    if offsets is None:
```
Comment (CCInc) [inline torchsparse/nn/functional/hash.py:21]: can we merge these two conditionals?

**Review by CCInc — APPROVED** [review]
Great! I love how clean the pytorch API is now.

File: torchsparse/backend/convolution/convolution_gpu.cu [hunk 9]
```diff
@@ -1,9 +1,10 @@
 #ifndef GPU_CONVOLUTION
```
[the remark this answers is not in the document]
Comment (CCInc) [reply torchsparse/backend/convolution/convolution_gpu.cu]: I think we can rename `convolution.cu` -> `convolution_gpu.cu` as well

[the remark this answers is not in the document]
Comment (zhijian-liu) [reply torchsparse/backend/convolution/convolution_gpu.cu]: Removed and renamed. Thanks!

File: torchsparse/nn/functional/hash.py (hunk 8 above)
Comment (zhijian-liu) [reply torchsparse/nn/functional/hash.py:17]: Merged.

File: torchsparse/nn/functional/devoxelize.py (hunk 7 above)
Comment (zhijian-liu) [reply torchsparse/nn/functional/devoxelize.py]: Added.

File: torchsparse/nn/functional/count.py (hunk 6 above)
Comment (zhijian-liu) [reply torchsparse/nn/functional/count.py]: Added.

File: torchsparse/nn/functional/conv.py (hunk 5 above)
Comment (zhijian-liu) [reply torchsparse/nn/functional/conv.py]: Added.

File: torchsparse/backend/voxelize/voxelize_gpu.h [hunk 10]
```diff
@@ -0,0 +1,14 @@
+#ifndef TORCHSPARSE_VOXELIZE_CUDA
+#define TORCHSPARSE_VOXELIZE_CUDA
+
+#include <torch/torch.h>
+
+#include <vector>
+
+at::Tensor voxelize_forward_cpu(const at::Tensor inputs, const at::Tensor idx,
```
Comment (kentangSJTU) [inline torchsparse/backend/voxelize/voxelize_gpu.h]: Here is a typo.

Comment (zhijian-liu) [reply torchsparse/backend/voxelize/voxelize_gpu.h]: Fixed.
